### ISSUE TITLE
[ETL] Incremental Pipeline

### ISSUE THREAD

**Issue by idiom-bytes** [oceanprotocol/pdr-backend#1001]
### Motivation

The ETL needs to be able to fetch 5 mins of data, every 5 minutes, in such a way that each fetch updates historical records allowing us to understand when previous predictions became correct, calculating revenue, accuracy, and many other metrics.

### Update Step - Incrementally updating the Lake
When you run the "lake update"
1. New records will arrive that need to be processed
1. These new records (such as pdr_payout) if applicable should be: (a) cleaned up into their raw/bronze table, (b) update historical records to reflect this event arriving such as a prediction that is now paid out.
1. After all records have been yielded to temp tables and the pipeline ends, records are then swapped to prod tables.

![Screenshot from 2024-05-06 13-27-25](https://github.com/oceanprotocol/pdr-backend/assets/69865342/23eb0e09-1fbe-40c4-88d5-be0aac13c1e6)

Data Workflows
- All data workflows will take place as a series of SQL queries, and different tables used to execute the incremental/swap strategy.
- All data workflows will have tests and coverage demonstrating it working.

### DoD:
- [x] Prediction and payouts are being processed incrementally
- [x] Bronze prediction is being updated incrementally
- [x] Other tables and bronze tables are currently not processed

Task:
- [x] Process new pdr-predictions into duckdb
- [x] Process new pdr-payouts into duckdb
- [x] Verify the incremental update step works [#982](https://github.com/oceanprotocol/pdr-backend/issues/982)
- [x] Create SQL that process new pdr-predictions into new_events_bronze_predictions_table
- [x] Create SQL that process new pdr-payouts into update_events_bronze_predictions_table
- [x] Create SQL & merge strategy that processes new_events and update_events into incremental bronze_tables
- [x] Verify that null records inside bronze-predictions are being updated correctly
- [x] Verify everything is working e2e

**Comment by idiom-bytes** [comment]
To implement this, we should first start w/ simply updating predictions when truevals and payouts show up 

[How this ticket grows]
- Basically 1 Event -> Multiple Table Inserts & Updates
- We need to consider what each event is doing, across all tables.

**subscription event**
-> new subscription record
-> new slot record

**prediction event**
-> new prediction record
-> update slot record

**trueval event**
-> new trueval record
-> update N prediction records
-> update 1 slot record

**payout event**
-> new payout record
-> update N prediction records
-> update 1 slot record

**Comment by idiom-bytes** [comment]
Here is one of my ways...
1. We update ETL so tables aren't attached to queries, it's just a set of queries
1. We then add multiple queries and another flow ("_update" tables) so we can reconcile everything.

### 1- process predictions
- We can simplify the query to just the insert logic, we don't need to join anymore
- We then process truevals
- We can just insert new ones to _bronze_truevals if we want to, but we're not doing that right now... So skip
- We want to create update events for the bronze_prediction table, so we extract the id, slot, trueval, and add it to the update_bronze_prediction

### 2- process payouts
- We can just insert new ones to _bronze_payouts if we want to, but we're not doing that right now... So skip
- We want to create update events for the bronze_prediction table, so we extract the id, slot, payout, and add it to the update_bronze_prediction

### 3 - Reduce updates
- We now have all update rows generated, and can begin the swap/finalize process.
- We then have the final step, where we "reduce", or join _etl_bronze_predictions with _update_bronze_predictions such that we can get all rows written to the final/live bronze_predictions table.

Although this takes a couple of extra steps, the overall amount of rows scanned/computed/joined, is far lower... Increasing the overall performance of the workflow.

Most of this work should look like SQL queries and a swap logic update at the end of the ETL update logic.

### 1 - Extract prediction update events from trueval 
![1_extract_truevals](https://github.com/oceanprotocol/pdr-backend/assets/69865342/8b7222a6-c35e-4b21-9023-9bacbb28a597)

### 2 - Extract prediction update events from payout
![2_extract_payouts](https://github.com/oceanprotocol/pdr-backend/assets/69865342/446e2586-124b-450a-8696-d1c835ed73d4)

### 3 - Prepare prediction updates and merge to final table
![3_update_to_final](https://github.com/oceanprotocol/pdr-backend/assets/69865342/1f10b013-660b-46f4-9c7c-23205e7da970)

### All prediction events per source
![4_all_prediction_events](https://github.com/oceanprotocol/pdr-backend/assets/69865342/07d388ab-f194-4c6d-9a8c-fe94be9fb599)

**Comment by idiom-bytes** [comment]
**[Feedback Mustafa]**
With reference to the code/design provided, as I explained to Mustafa after reviewing his proposal.
- doing an onEventHandler approach will likely lead to more scans/compute than required
- the way it's recommended, it looks to be doing more copies than needed

**[Effective Processing of Events]**
I have instead, done a pseudo-implementation of the SQL queries + logic required to get this working.
- It only reads from each raw tables once (small, incoming tables)
- It only joins with each bronze_table once (large historical tables)
- rather than joining all update records, it tries to group/reduce them together to have only 1:1 join with the large historical table

**[Simplify Requirements even Further]**
I have also emphasized **how much simpler all of this can be to deliver on the goal of: predictoor revenue dashboard** by not requiring the trueval table.

Trueval does not contain the user id anywhere, so it cannot update the prediction table directly.
First, it will need to update a slot, **and we're not caring about that at the moment.**

Literally, all we need is to join payouts with predictions.
The rest can come later.
